core/src/main/java/brooklyn/config/internal/AbstractConfigMapImpl.java

```
@@ -0,0 +1,110 @@
+    
+    /**
+     * Map of configuration information that is defined at start-up time for the entity. These
+     * configuration parameters are shared and made accessible to the "children" of this
+     * entity.
+     */
+    protected Map<ConfigKey<?>,Object> ownConfig = Collections.synchronizedMap(new LinkedHashMap<ConfigKey<?>, Object>());
+
+    public <T> T getConfig(ConfigKey<T> key) {
+        return getConfig(key, null);
+    }
+    
+    public <T> T getConfig(HasConfigKey<T> key) {
+        return getConfig(key.getConfigKey(), null);
+    }
+    
+    public <T> T getConfig(HasConfigKey<T> key, T defaultValue) {
+        return getConfig(key.getConfigKey(), defaultValue);
+    }
+    
+    @Override @Deprecated
+    public Object getRawConfig(ConfigKey<?> key) {
+        return getConfigRaw(key, true).orNull();
+    }
+    
+    protected Object setConfigPrep1(ConfigKey<?> key, Object v) {
+        Object val;
+        if ((v instanceof Future) || (v instanceof DeferredSupplier)) {
+            // no coercion for these (coerce on exit)
+            val = v;
+        } else if (key instanceof StructuredConfigKey) {
+            // no coercion for these structures (they decide what to do)
+            val = v;
+        } else if ((v instanceof Map || v instanceof Iterable) && key.getType().isInstance(v)) {
+            // don't do coercion on put for these, if the key type is compatible, 
+            // because that will force resolution deeply
+            val = v;
+        } else {
+            try {
+                // try to coerce on input, to detect errors sooner
+                val = TypeCoercions.coerce(v, key.getTypeToken());
+            } catch (Exception e) {
+                throw new IllegalArgumentException("Cannot coerce or set "+v+" to "+key, e);
+                // if can't coerce, we could just log, and *throw* the error when we retrieve the config
```

Include a `TODO` in the comment for things that are to be changed in the future. Can you include in the comment why for now we are failing fast, so when would someone want to change this.